README.md
# Pre Charts (Responsive)

![Change Log](./CHANGELOG.md)

## Installation

```
npm install --save pre-charts
```

## AreaChart

![alt tag](./area.png)

```js
import AreaChart from "pre-charts/lib/AreaChart";
...

let data = [
    {name: "A", public: 4000, private: 2400, protected: 2400},
    {name: "B", public: 3000, private: 1398, protected: 2210},
    {name: "C", public: 2000, private: 9800, protected: 2290},
    ...
];

let meta = [
    {dataKey: "public", name: "Public", unit: "piece", fill: "#F44336"},
    {dataKey: "private", name: "Private", unit: "piece", fill: "red"},
    {dataKey: "protected", name: "Protected", unit: "piece", fill: "blue"}
];
...

<AreaChart label={"Area Chart"} data={data} width={400} height={250} meta={meta}/>

```

## BarChart

![alt tag](./bar.png)

```js
import BarChart from "pre-charts/lib/BarChart";
...

let data = [
    {name: "A", public: 4000, private: 2400, protected: 2400},
    {name: "B", public: 3000, private: 1398, protected: 2210},
    {name: "C", public: 2000, private: 9800, protected: 2290},
    ...
];

let meta = [
    {dataKey: "public", name: "Public", unit: "piece", fill: "blue"},
    {dataKey: "private", name: "Private", unit: "piece", fill:"#F44336"},
    {dataKey: "protected", name: "Protected", unit: "piece", fill: "red"}
];
...

<BarChart label={"Bar Chart"} data={data} width={400} height={250} meta={meta}/>

```

## LineChart

![alt tag](./line.png)

```js
import LineChart from "pre-charts/lib/LineChart";
...

let data = [
    {name: "A", public: 4000, private: 2400, protected: 2400},
    {name: "B", public: 3000, private: 1398, protected: 2210},
    {name: "C", public: 2000, private: 9800, protected: 2290},
    ...
];

let meta = [
    {dataKey: "public", name: "Public", unit: "piece", fill:"red"},
    {dataKey: "private", name: "Private", unit: "piece", fill:"blue"},
    {dataKey: "protected", name: "Protected", unit: "piece", fill:"yellow"}
];
...

<LineChart label={"Line Chart"} data={data} width={400} height={250} meta={meta}/>

```


## ScatterChart

![alt tag](./scatter.png)


```js
import ScatterChart from "pre-charts/lib/ScatterChart";
...

const dataA = [
    {x: 100, y: 200, z: 200},
    {x: 120, y: 100, z: 260},
    {x: 170, y: 300, z: 400},
    {x: 140, y: 250, z: 280},
    {x: 150, y: 400, z: 500},
    {x: 110, y: 280, z: 200},
    {x: 100, y: 200, z: 200},
    {x: 120, y: 100, z: 260},
    {x: 170, y: 300, z: 400},
    {x: 140, y: 250, z: 280},
    {x: 150, y: 400, z: 500},
    {x: 110, y: 280, z: 200}
];

const dataB = [
    {x: 200, y: 260, z: 240},
    {x: 240, y: 290, z: 220},
    {x: 190, y: 290, z: 250},
    {x: 198, y: 250, z: 210},
    {x: 180, y: 280, z: 260},
    {x: 210, y: 220, z: 230}];

const data = [
    {name: "A", data: dataA},
    {name: "B", data: dataB}
];

let meta = [
    {dataKey: "x", unit: "cm", name: "X", fill:"#F44336"},
    {dataKey: "y", unit: "cm", name: "Y", fill:"red"},
    {dataKey: "z", unit: "cm", name: "Z", fill:"blue"}
];

...

<ScatterChart label={"Scatter Chart"} data={data} width={400} height={250} meta={meta}/>

```


## PieChart

![alt tag](./pie.png)

```js
import PieChart from "pre-charts/lib/PieChart";
...

let data = [
    {
        value: 1500,
        label: "A",
        key: "0",
        unit: "ms",
        fill:"#2196F3",
        children:[
                {
                    value: 2500,
                    label: "A1",
                    key: "11",
                    unit: "ms",
                    fill:"#F44336"
                },
                {
                    value: 3000,
                    label: "A2",
                    key: "12",
                    unit: "ms",
                    fill:"red"
                }
        ]
    },
    {
        value: 2500,
        label: "B",
        key: "1",
        unit: "ms",
        fill:"#F44336"
    },
    {
        value: 3000,
        label: "C",
        key: "3",
        unit: "ms",
        fill:"red"
    }
];

...

<PieChart label={"Pie Chart"} size={200} data={data}/>

```

### Quick Start

#### 1. Get the latest version
You can start by cloning the latest version of pre-charts.

#### 2. Run `npm install`
This will install both run-time project dependencies and developer tools listed
in [package.json](./package.json) file.

#### 3. How to start project in Development Mode

This will start the development server and serve site application.

```shell
$ npm start
```
  
Open Browser and enter `http://localhost:3000` (default) 

### How to Build for Production

If you need just to build the app (without running a dev server), simply run:

```shell
$ npm run build
```

### How to Build for Site

If you need just to build the app (without running a dev server), simply run:

```shell
$ npm run site
```


CHANGELOG.md
# Change Log

## 1.0.2 Developing...

## 1.0.1
* Added responsive support


lib/LineChart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _ResizeComponent = require("./ResizeComponent");

var _ResizeComponent2 = _interopRequireDefault(_ResizeComponent);

var _Legend = require("./Legend");

var _Legend2 = _interopRequireDefault(_Legend);

var _Arrays = require("./utils/Arrays");

var _Arrays2 = _interopRequireDefault(_Arrays);

require("./LineChart.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LineChart = function (_Component) {
    _inherits(LineChart, _Component);

    function LineChart(props) {
        _classCallCheck(this, LineChart);

        var _this = _possibleConstructorReturn(this, (LineChart.__proto__ || Object.getPrototypeOf(LineChart)).call(this, props));

        _this.legends = [];

        _this.__onResize = function (_ref) {
            var width = _ref.width,
                height = _ref.height;

            var scale = Math.max(width, _this.props.minWidth) / (_this.props.width + 40);
            var margin = 100 - scale * 100;
            _this.setState({
                scale: scale <= 1 ? scale : 1,
                marginTop: margin > 0 ? -1 * (_this.props.height / 100 + 1) * margin : 0
            });
        };

        _this.__renderLines = function (data, meta) {
            var metaArr = [];
            var xAxisWidth = _this.__xAxisWidth();
            var sumXAxisWidth = 0;

            for (var i = 0; i < data.length - 1; i++) {
                var item = data[i];
                var itemArr = [];
                var nexItem = void 0;

                if (data[i + 1]) {
                    nexItem = data[i + 1];
                }

                var fields = _this.__getFields(item);

                for (var j in fields) {

                    var key = fields[j].key,
                        value = fields[j].value,
                        nexValue = value;

                    if (nexItem) {
                        nexValue = nexItem[key];
                    }

                    var properties = _Arrays2.default.getValueByKey(meta, "dataKey", key);
                    properties = properties === undefined ? {} : properties;

                    var tooltip = item.name + "\n" + (properties.name || key) + " : " + value + " " + (properties.unit || "") + "\n",
                        tooltipNext = nexItem.name + "\n" + (properties.name || key) + " : " + nexValue + " " + (properties.unit || "") + "\n";

                    var fill = properties.fill || _this.__randColor(j);

                    _this.legends[properties.name || key] = {fill: fill, label: properties.name || key};

                    var pointY = _this.__pointY(value),
                        nextPointY = _this.__pointY(nexValue),
                        x1 = sumXAxisWidth,
                        y1 = pointY,
                        x2 = sumXAxisWidth + xAxisWidth,
                        y2 = nextPointY;

                    itemArr.push(_react2.default.createElement("line", {
                        key: key,
                        x1: x1,
                        y1: y1,
                        x2: x2,
                        y2: y2,
                        data: tooltip + tooltipNext,
                        strokeLinecap: "round",
                        stroke: fill,
                        strokeWidth: "4" }));
                }

                sumXAxisWidth += xAxisWidth;

                metaArr.push(_react2.default.createElement(
                    "g",
                    { key: i },
                    itemArr
                ));
            }
            return metaArr;
        };

        _this.__renderYAxis = function () {
            var maxYAxis = _this.__maxYAxis();
            var height = _this.props.height / 4;
            var axisArr = [];
            for (var i = 0; i < 4; i++) {
                axisArr.push(_react2.default.createElement("div", { key: i,
                    id: parseInt(maxYAxis / 4 * (4 - i)),
                    className: "rb-line-y-axis",
                    style: { height: height } }));
            }
            return axisArr;
        };

        _this.__renderXAxis = function () {
            var data = _this.props.data;
            var width = _this.__xAxisWidth();
            var axisArr = [];
            for (var i = 0; i < data.length - 1; i++) {
                var item = data[i],
                    nextItem = data[i + 1] || {};
                axisArr.push(_react2.default.createElement("div", { key: i,
                    className: "rb-line-x-axis",
                    data: item.name,
                    name: nextItem.name,
                    style: { width: width } }));
            }
            return axisArr;
        };

        _this.__maxYAxis = function () {
            var data = _this.props.data;
            var maxYAxis = 0;
            for (var i in data) {
                var fields = _this.__getFields(data[i]);
                for (var j in fields) {
                    if (fields[j].value > maxYAxis) maxYAxis = fields[j].value;
                }
            }
            var a = maxYAxis > 1000 ? 1000 : maxYAxis > 100 ? 100 : maxYAxis > 50 ? 50 : maxYAxis > 10 ? 10 : 1;

            return ~~((maxYAxis + a - 1) / a) * a;
        };

        _this.__xAxisWidth = function () {
            return (_this.props.width - 1) / (_this.props.data.length - 1);
        };

        _this.__pointY = function (value) {
            var maxYAxis = _this.__maxYAxis();
            return _this.props.height * (value * 100 / maxYAxis) / 100;
        };

        _this.__getFields = function (data) {
            var arr = [];
            for (var key in data) {
                if (data.hasOwnProperty(key)) {
                    if (key === "name" || key === "fill" || key === "unit") {
                        continue;
                    }
                    arr.push({
                        value: data[key],
                        key: key
                    });
                }
            }
            return arr;
        };

        _this.__randColor = function (index) {
            var colors = ["#F44336", "#673AB7", "#2196F3", "#FF5722", "#9C27B0", "#FFC107", "#FF9800", "#4CAF50", "#00796B", "#009688", "#3F51B5"];
            if (index !== undefined) {
                return colors[index % colors.length];
            }
            return colors[Math.floor(Math.random() * (colors.length - 1))];
        };

        _this.state = {
            scale: 1,
            marginTop: 0
        };
        return _this;
    }

    _createClass(LineChart, [{
        key: "render",
        value: function render() {
            var label = this.props.label === undefined ? undefined : _react2.default.createElement(
                    "h3",
                    {className: "rb-line-chart-label"},
                    this.props.label
                );
            return _react2.default.createElement(
                _ResizeComponent2.default,
                {
                    style: {
                        transform: "scale(" + this.state.scale + ")",
                        transformOrigin: "bottom left",
                        marginTop: this.state.marginTop
                    },
                    onResize: this.__onResize
                },
                label,
                _react2.default.createElement(
                    "div",
                    {className: "rb-line-chart", style: {width: this.props.width, height: this.props.height}},
                    _react2.default.createElement(
                        "svg",
                        {className: "rb-line-chart-svg"},
                        this.__renderLines(this.props.data, this.props.meta)
                    ),
                    _react2.default.createElement(
                        "div",
                        {className: "rb-line-chart-axis"},
                        this.__renderYAxis()
                    ),
                    _react2.default.createElement(
                        "div",
                        {className: "rb-line-chart-axis"},
                        this.__renderXAxis()
                    )
                ),
                _react2.default.createElement(_Legend2.default, {
                    data: this.legends,
                    width: this.props.legendWidth || this.props.width
                })
            );
        }
    }]);

    return LineChart;
}(_react.Component);

LineChart.propTypes = {
    /**
     * Label for the form control.
     */
    label: _react2.default.PropTypes.any,
    /**
     * Width for chart as px
     */
    width: _react2.default.PropTypes.number,
    /**
     * Min width for chart as px
     */
    minWidth: _react2.default.PropTypes.number,
    /**
     * Height for chart as px
     */
    height: _react2.default.PropTypes.number,
    /**
     * Data to be plotted on the chart
     */
    data: _react2.default.PropTypes.array,
    /**
     * Change to be made for the given data
     */
    meta: _react2.default.PropTypes.array,
    /**
     *
     */
    legendWidth: _react2.default.PropTypes.number
};
LineChart.defaultProps = {
    width: 500,
    height: 300,
    minWidth: 150,
    data: [],
    meta: []
};
exports.default = LineChart;